File: singleton/java/Singleton.java
// Singleton class: Database
class Database {
    // Static instance variable (volatile for thread safety)
    private static volatile Database instance;
    private String connection;

    // Private constructor to prevent instantiation
    private Database() {
        this.connection = "Database connection established";
        System.out.println("Creating database instance");
    }

    // Thread-safe getInstance method using double-checked locking
    public static Database getInstance() {
        if (instance == null) {
            synchronized (Database.class) {
                if (instance == null) {
                    instance = new Database();
                }
            }
        }
        return instance;
    }

    public String query(String query) {
        return "Executing query: " + query + " on " + connection;
    }
}

// Main class demonstrating Singleton pattern
public class Singleton {
    public static void main(String[] args) {
        // Get singleton instance
        Database db1 = Database.getInstance();
        System.out.println(db1.query("SELECT * FROM users"));

        System.out.println();

        // Get singleton instance again (same instance)
        Database db2 = Database.getInstance();
        System.out.println(db2.query("SELECT * FROM products"));

        System.out.println();

        // Verify both are the same instance
        System.out.println("db1 and db2 are the same instance: " + (db1 == db2));
    }
}

File: patterns/singleton/python/Singleton.py
import threading

# Singleton class: Database
class Database:
    _instance = None
    _lock = threading.Lock()

    # Private constructor logic via __new__
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance.connection = "Database connection established"
                    print("Creating database instance")
        return cls._instance

    def query(self, sql: str) -> str:
        return f"Executing query: {sql} on {self.connection}"


# Main: demonstrating Singleton pattern
if __name__ == "__main__":
    # Get singleton instance
    db1 = Database()
    print(db1.query("SELECT * FROM users"))

    print()

    # Get singleton instance again (same instance)
    db2 = Database()
    print(db2.query("SELECT * FROM products"))

    print()

    # Verify both are the same instance
    print(f"db1 and db2 are the same instance: {db1 is db2}")
